lib/cassandra/cluster/schema/fetchers.rb

```diff
@@ -284,12 +284,24 @@ def create_keyspace(keyspace_data, rows_tables, rows_columns,
             types = rows_types.each_with_object({}) do |row, types|
                       types[row['type_name']] = create_type(row)
                     end
-            functions = rows_functions.each_with_object({}) do |row, functions|
-                          functions[row['function_name']] = create_function(row)
-                        end
-            aggregates = rows_aggregates.each_with_object({}) do |row, aggregates|
-                           aggregates[row['aggregate_name']] = create_aggregate(row, functions)
-                         end
+
+            # This is a somewhat fancy (though I believe efficient) way of making a hash
+            # from an array of key-value elements. We want the functions hash to be keyed
+            # on [name, arg-types-list]. Similarly for the aggregates hash.
+
+            functions = Hash[
+                rows_functions.map do |row|
+                  func = create_function(row)
+                  [[func.name, func.argument_types], func]
+                end
+            ]
```

Comment:
I don't think I understand the performance advantage argument here either. We still map (iterate) over an entire array in pure ruby, and then iterate it again in C, so instead of a single slower iteration, we get a slower iteration and a faster iteration, which is still slower. Not even considering all the temporary variables we're creating.